// node-syphon/dist/node-syphon-universal.mjs
const SyphonServerDescriptionNameKey = "SyphonServerDescriptionNameKey";
const SyphonServerDescriptionAppNameKey = "SyphonServerDescriptionAppNameKey";
const SyphonServerDescriptionUUIDKey = "SyphonServerDescriptionUUIDKey";
var SyphonServerDirectoryListenerChannel = /* @__PURE__ */ ((SyphonServerDirectoryListenerChannel2) => {
  SyphonServerDirectoryListenerChannel2["SyphonServerInfoNotification"] = "info";
  SyphonServerDirectoryListenerChannel2["SyphonServerErrorNotification"] = "error";
  SyphonServerDirectoryListenerChannel2["SyphonServerAnnounceNotification"] = "announce";
  SyphonServerDirectoryListenerChannel2["SyphonServerRetireNotification"] = "retire";
  SyphonServerDirectoryListenerChannel2["SyphonServerUpdateNotification"] = "update";
  return SyphonServerDirectoryListenerChannel2;
})(SyphonServerDirectoryListenerChannel || {});
export {
  SyphonServerDescriptionAppNameKey,
  SyphonServerDescriptionNameKey,
  SyphonServerDescriptionUUIDKey,
  SyphonServerDirectoryListenerChannel
};
