// node-syphon/dist/node-syphon.mjs
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
import bindings from "bindings";
import { fork } from "child_process";
import { dirname, resolve } from "path";
import { fileURLToPath } from "url";
const SyphonServerDescriptionNameKey = "SyphonServerDescriptionNameKey";
const SyphonServerDescriptionAppNameKey = "SyphonServerDescriptionAppNameKey";
const SyphonServerDescriptionUUIDKey = "SyphonServerDescriptionUUIDKey";
const SyphonAddon = bindings({
  bindings: "syphon",
  // The bin folder from the lib one.
  try: [
    // For standard installation.
    ["module_root", "node_modules", "node-syphon", "dist", "bin", "syphon.node"],
    // For local examples (monorepo).
    ["module_root", "dist", "bin", "syphon.node"]
  ]
});
class SyphonOpenGLServer {
  constructor(name) {
    __publicField(this, "_server");
    __publicField(this, "_dataDeprecationWarned", false);
    __publicField(this, "_handleDeprecationWarned", false);
    this._server = new SyphonAddon.OpenGLServer(name);
  }
  dispose() {
    this._server.dispose();
  }
  publishImageData(data, imageRegionOrTextureTarget, textureDimensionOrImageRegion, flippedOrTextureDimension, textureTargetOrFlipped) {
    if (typeof imageRegionOrTextureTarget === "string") {
      if (process && process.emitWarning && !this._dataDeprecationWarned) {
        process.emitWarning(
          "SyphonOpenGLServer.publishImageData(data, textureTarget, imageRegion, textureDimension, flipped) is deprecated. Use publishImageData(data, imageRegion, textureDimension, flipped, textureTarget?) instead.",
          "DeprecationWarning"
        );
        this._dataDeprecationWarned = true;
      }
      this._server.publishImageData(
        data,
        textureDimensionOrImageRegion,
        flippedOrTextureDimension,
        textureTargetOrFlipped,
        imageRegionOrTextureTarget
      );
      return;
    }
    this._server.publishImageData(
      data,
      imageRegionOrTextureTarget,
      textureDimensionOrImageRegion,
      flippedOrTextureDimension,
      typeof textureTargetOrFlipped === "string" ? textureTargetOrFlipped : "GL_TEXTURE_RECTANGLE_EXT"
    );
  }
  publishSurfaceHandle(handle, imageRegionOrTextureTarget, textureDimensionOrImageRegion, flippedOrTextureDimension, textureTargetOrFlipped) {
    if (typeof imageRegionOrTextureTarget === "string") {
      if (process && process.emitWarning && this._handleDeprecationWarned) {
        process.emitWarning(
          "SyphonOpenGLServer.publishSurfaceHandle(handle, textureTarget, imageRegion, textureDimension, flipped) is deprecated. Use publishSurfaceHandle(handle, imageRegion, textureDimension, flipped, textureTarget?) instead.",
          "DeprecationWarning"
        );
        this._handleDeprecationWarned = true;
      }
      this._server.publishSurfaceHandle(
        handle,
        textureDimensionOrImageRegion,
        flippedOrTextureDimension,
        textureTargetOrFlipped,
        imageRegionOrTextureTarget
      );
      return;
    }
    this._server.publishSurfaceHandle(
      handle,
      imageRegionOrTextureTarget,
      textureDimensionOrImageRegion,
      flippedOrTextureDimension,
      typeof textureTargetOrFlipped === "string" ? textureTargetOrFlipped : "GL_TEXTURE_RECTANGLE_EXT"
    );
  }
  /**
   * Publish a raw OpenGL texture ID.
   * This is the most efficient method for cables.gl as it stays entirely on the GPU.
   */
  publish(textureId, width, height, textureTarget = "GL_TEXTURE_2D", flipped = false) {
    this._server.publish(textureId, width, height, textureTarget, flipped);
  }
  /**
   * Publish the currently bound texture.
   * Useful when the raw texture ID is not directly accessible (e.g. WebGLTexture objects).
   */
  publishBoundTexture(width, height, textureTarget = "GL_TEXTURE_2D", flipped = false) {
    this._server.publishBoundTexture(width, height, textureTarget, flipped);
  }
  get name() {
    return this._server.name;
  }
  get description() {
    return this._server.serverDescription;
  }
  get hasClients() {
    return this._server.hasClients;
  }
}
class SyphonOpenGLClient {
  constructor(description) {
    __publicField(this, "client");
    __publicField(this, "listeners", []);
    __publicField(this, "isFrameListenerSet", false);
    this.client = new SyphonAddon.OpenGLClient(description);
  }
  dispose() {
    this.listeners.length = 0;
    this.isFrameListenerSet = false;
    this.client.dispose();
  }
  on(channel, callback) {
    switch (channel) {
      case "frame": {
        process.emitWarning(
          `SyphonOpenGLClient.on('frame') is deprecated. Use on('data') instead.`,
          "DeprecationWarning"
        );
      }
      case "data": {
        if (!this.isFrameListenerSet) {
          this.client.on("frame", this.frameDataListenerCallback.bind(this));
          this.isFrameListenerSet = true;
        }
        this.listeners.push(callback);
        break;
      }
    }
  }
  off(channel, callback) {
    switch (channel) {
      case "frame": {
        process.emitWarning(
          `SyphonOpenGLClient.off('frame') is deprecated. Use off('data') instead.`,
          "DeprecationWarning"
        );
      }
      case "data": {
        const index = this.listeners.indexOf(callback);
        if (index >= 0) {
          this.listeners.splice(index, 1);
          if (this.listeners.length === 0) {
            this.client.off("frame");
            this.isFrameListenerSet = false;
          }
        }
        break;
      }
    }
  }
  frameDataListenerCallback(frame) {
    for (const listener of this.listeners) {
      listener(frame);
    }
  }
}
class SyphonMetalServer {
  constructor(name) {
    __publicField(this, "_server");
    __publicField(this, "_dataDeprecationWarned", false);
    this._server = new SyphonAddon.MetalServer(name);
  }
  dispose() {
    this._server.dispose();
  }
  publishImageData(data, imageRegion, textureDimensionOrBytesPerRow, flipped) {
    if (typeof textureDimensionOrBytesPerRow === "number") {
      if (process && process.emitWarning && !this._dataDeprecationWarned) {
        process.emitWarning(
          "SyphonMetalServer.publishImageData(data, imageRegion, bytePerRow, flipped) is deprecated. Use publishImageData(data, imageRegion, textureDimension, flipped) instead.",
          "DeprecationWarning"
        );
        this._dataDeprecationWarned = true;
      }
      this._server.publishImageData(
        data,
        imageRegion,
        { width: imageRegion.width, height: imageRegion.height },
        flipped
      );
      return;
    }
    this._server.publishImageData(data, imageRegion, textureDimensionOrBytesPerRow, flipped);
  }
  publishSurfaceHandle(handle, imageRegion, textureDimension, flipped) {
    this._server.publishSurfaceHandle(handle, imageRegion, textureDimension, flipped);
  }
  get name() {
    return this._server.name;
  }
  get description() {
    return this._server.serverDescription;
  }
  get hasClients() {
    return this._server.hasClients;
  }
}
class SyphonMetalClient {
  constructor(description) {
    __publicField(this, "client");
    __publicField(this, "frameListeners", []);
    __publicField(this, "textureListeners", []);
    __publicField(this, "isFrameListenerSet", false);
    __publicField(this, "isTextureListenerSet", false);
    this.client = new SyphonAddon.MetalClient(description);
  }
  dispose() {
    this.frameListeners.length = 0;
    this.isFrameListenerSet = false;
    this.textureListeners.length = 0;
    this.isTextureListenerSet = false;
    this.client.dispose();
  }
  on(channel, callback) {
    switch (channel) {
      case "frame": {
        process.emitWarning(
          `SyphonMetalClient.on('frame') is deprecated. Use on('data') instead.`,
          "DeprecationWarning"
        );
      }
      case "data": {
        if (!this.isFrameListenerSet) {
          this.client.on("frame", this.frameDataListenerCallback.bind(this));
          this.isFrameListenerSet = true;
        }
        this.frameListeners.push(callback);
        break;
      }
      case "texture": {
        if (!this.isTextureListenerSet) {
          this.client.on("texture", this.textureHandleListenerCallback.bind(this));
          this.isTextureListenerSet = true;
        }
        this.textureListeners.push(callback);
        break;
      }
    }
  }
  off(channel, callback) {
    switch (channel) {
      case "frame": {
        const index = this.frameListeners.indexOf(callback);
        if (index >= 0) {
          this.frameListeners.splice(index, 1);
          if (this.frameListeners.length === 0) {
            this.client.off("frame");
            this.isFrameListenerSet = false;
          }
        }
        break;
      }
      case "texture": {
        const index = this.textureListeners.indexOf(callback);
        if (index >= 0) {
          this.textureListeners.splice(index, 1);
          if (this.textureListeners.length === 0) {
            this.client.off("texture");
            this.isTextureListenerSet = false;
          }
        }
        break;
      }
    }
  }
  frameDataListenerCallback(frame) {
    for (const listener of this.frameListeners) {
      listener(frame);
    }
  }
  textureHandleListenerCallback(frame) {
    for (const listener of this.textureListeners) {
      listener({
        ...frame,
        release: () => {
          this.client.releaseTexture(frame.frameCount);
        }
      });
    }
  }
}
var SyphonServerDirectoryListenerChannel = /* @__PURE__ */ ((SyphonServerDirectoryListenerChannel2) => {
  SyphonServerDirectoryListenerChannel2["SyphonServerInfoNotification"] = "info";
  SyphonServerDirectoryListenerChannel2["SyphonServerErrorNotification"] = "error";
  SyphonServerDirectoryListenerChannel2["SyphonServerAnnounceNotification"] = "announce";
  SyphonServerDirectoryListenerChannel2["SyphonServerRetireNotification"] = "retire";
  SyphonServerDirectoryListenerChannel2["SyphonServerUpdateNotification"] = "update";
  return SyphonServerDirectoryListenerChannel2;
})(SyphonServerDirectoryListenerChannel || {});
const __dirname = dirname(fileURLToPath(import.meta.url));
const PATH = `${resolve(__dirname, "server-directory-process.js")}`;
const EXIT_TYPES = [
  `exit`,
  `SIGINT`,
  `SIGUSR1`,
  `SIGUSR2`,
  `SIGQUIT`,
  `SIGSEGV`,
  `uncaughtException`,
  `SIGTERM`
];
const NodeSyphonMessageTypeKey = "NodeSyphonMessageType";
const NodeSyphonMessageKey = "NodeSyphonMessage";
const NodeSyphonMessageTypeInfo = "NodeSyphonMessageInfo";
const NodeSyphonMessageTypeError = "NodeSyphonMessageError";
const NodeSyphonMessageTypeNotification = "NodeSyphonMessageNotification";
const NodeSyphonNotificationTypeKey = "NodeSyphonNotificationType";
const NodeSyphonServerDictionaryKey = "NodeSyphonServerDictionary";
const _SyphonServerDirectory = class _SyphonServerDirectory {
  constructor() {
    __publicField(this, "directoryProcess");
    __publicField(this, "directoryRunning", false);
    /**
     * Current servers in directory.
     */
    __publicField(this, "_servers", []);
    /**
     * Listeners on servers' notifications.
     */
    __publicField(this, "_listeners", {});
    if (_SyphonServerDirectory.instance) {
      return _SyphonServerDirectory.instance;
    }
    _SyphonServerDirectory.instance = this;
  }
  dispose() {
    this.removeAllListeners();
    if (this.directoryProcess) {
      this.directoryProcess.kill();
      this.directoryProcess = null;
      this._servers.length = 0;
      this.directoryRunning = false;
    }
  }
  on(channel, fn) {
    if (!this._listeners[channel]) {
      this._listeners[channel] = [];
    }
    this._listeners[channel].push(fn);
  }
  off(channel, fn) {
    if (this._listeners[channel]) {
      const callback = this._listeners[channel].find((listener) => listener == fn);
      if (callback) {
        const index = this._listeners[channel].indexOf(callback);
        this._listeners[channel].splice(index, 1);
      }
    }
  }
  removeAllListeners() {
    for (const channel in this._listeners) {
      this._listeners[channel].length = 0;
    }
  }
  get isRunning() {
    return this.directoryRunning;
  }
  get servers() {
    return this._servers;
  }
  get debug() {
    return {
      path: PATH,
      pid: this.directoryProcess.pid
    };
  }
  /**
   * Listens to servers' directory changes.
   */
  listen() {
    try {
      if (this.directoryRunning) {
        return;
      }
      this.handleExitSignal();
      this.directoryProcess = fork(PATH, {
        silent: true
      });
      this._emit(
        SyphonServerDirectoryListenerChannel.SyphonServerInfoNotification,
        `Syphon directory server process launched with pid: ${this.directoryProcess.pid}`
      );
      this.directoryProcess.stdout.setEncoding("utf8");
      this.directoryProcess.stdout.on("data", this._parseProcessData.bind(this));
      this.directoryProcess.stderr.setEncoding("utf8");
      this.directoryProcess.stderr.on("data", (err) => {
        throw new Error(err.toString());
      });
      this._emit(
        SyphonServerDirectoryListenerChannel.SyphonServerInfoNotification,
        `SyphonServerDirectory will set 'running: true`
      );
      this.directoryRunning = true;
    } catch (err) {
      this._emit(SyphonServerDirectoryListenerChannel.SyphonServerErrorNotification, err);
      this.dispose();
      console.error(err);
      throw err;
    }
  }
  _emit(channel, value) {
    if (this._listeners[channel]) {
      for (const callback of this._listeners[channel]) {
        callback(value);
      }
    }
  }
  _parseProcessData(data) {
    try {
      const split = data.toString().split(/\r?\n/);
      let obj = [];
      for (const str of split) {
        if (str.trim().length === 0)
          continue;
        obj.push(JSON.parse(str));
      }
      for (const message of obj) {
        switch (message.NodeSyphonMessageType) {
          case NodeSyphonMessageTypeInfo: {
            this._emit(
              SyphonServerDirectoryListenerChannel.SyphonServerInfoNotification,
              message.NodeSyphonMessage
            );
            break;
          }
          case NodeSyphonMessageTypeError: {
            this._emit(
              SyphonServerDirectoryListenerChannel.SyphonServerErrorNotification,
              message.NodeSyphonMessage
            );
            break;
          }
          case NodeSyphonMessageTypeNotification: {
            const type = message.NodeSyphonMessage.NodeSyphonNotificationType;
            switch (type) {
              case "SyphonServerAnnounceNotification": {
                this._servers.push(message.NodeSyphonMessage.NodeSyphonServerDictionary);
                this._servers = Array.from(new Set(this._servers));
                this._emit(
                  SyphonServerDirectoryListenerChannel.SyphonServerAnnounceNotification,
                  message.NodeSyphonMessage.NodeSyphonServerDictionary
                );
                break;
              }
              case "SyphonServerRetireNotification": {
                const server = this._servers.find(
                  (s) => s.SyphonServerDescriptionUUIDKey == message.NodeSyphonMessage.NodeSyphonServerDictionary.SyphonServerDescriptionUUIDKey
                );
                if (server) {
                  const index = this._servers.indexOf(server);
                  this._servers.splice(index, 1);
                  this._emit(
                    SyphonServerDirectoryListenerChannel.SyphonServerRetireNotification,
                    message.NodeSyphonMessage.NodeSyphonServerDictionary
                  );
                }
                break;
              }
              case "SyphonServerUpdateNotification": {
                const server = this._servers.find(
                  (s) => s.SyphonServerDescriptionUUIDKey == message.NodeSyphonMessage.NodeSyphonServerDictionary.SyphonServerDescriptionUUIDKey
                );
                if (server) {
                  const index = this._servers.indexOf(server);
                  this._servers.splice(
                    index,
                    1,
                    message.NodeSyphonMessage.NodeSyphonServerDictionary
                  );
                  this._emit(
                    SyphonServerDirectoryListenerChannel.SyphonServerUpdateNotification,
                    message.NodeSyphonMessage.NodeSyphonServerDictionary
                  );
                }
                break;
              }
            }
            break;
          }
          default: {
            throw new Error(
              `Unhandled message type from ServerDirectory subprocess: ${message.NodeSyphonMessageType}`
            );
          }
        }
      }
    } catch (err) {
      console.error(err);
      throw err;
    }
  }
  handleExitSignal() {
    process.stdin.resume();
    EXIT_TYPES.forEach((eventType) => {
      const end = () => {
        if (this.directoryProcess) {
          console.log("SyphonServerDirectory exits with event type:", eventType);
          this.dispose();
          EXIT_TYPES.forEach((eventType2) => {
            process.off(eventType2, end);
          });
        }
      };
      process.on(eventType, end);
    });
  }
};
__publicField(_SyphonServerDirectory, "instance");
let SyphonServerDirectory = _SyphonServerDirectory;
export {
  NodeSyphonMessageKey,
  NodeSyphonMessageTypeError,
  NodeSyphonMessageTypeInfo,
  NodeSyphonMessageTypeKey,
  NodeSyphonMessageTypeNotification,
  NodeSyphonNotificationTypeKey,
  NodeSyphonServerDictionaryKey,
  SyphonMetalClient,
  SyphonMetalServer,
  SyphonOpenGLClient,
  SyphonOpenGLServer,
  SyphonServerDescriptionAppNameKey,
  SyphonServerDescriptionNameKey,
  SyphonServerDescriptionUUIDKey,
  SyphonServerDirectory,
  SyphonServerDirectoryListenerChannel
};
